File: frontend/src/types/desolvation.ts
/**
 * Desolvation energy calculation types
 * 去溶剂化能计算相关类型定义
 */

/**
 * 去溶剂化模式
 * - stepwise: 逐级去溶剂（每次去掉一个配体）
 * - full: 全部去溶剂（直接计算中心离子单独能量）
 */
export type DesolvationMode = 'stepwise' | 'full';

/**
 * 溶剂模型类型
 * - gas: 气相（无溶剂效应）
 * - pcm: PCM 极化连续介质模型
 * - smd: SMD 溶剂密度模型
 * - custom: 自定义溶剂参数
 */
export type SolventModel = 'gas' | 'pcm' | 'smd' | 'custom';

/**
 * 溶剂配置
 */
export interface SolventConfig {
  model: SolventModel;
  solvent_name?: string;  // 预定义溶剂名称（如 water, acetonitrile）
  // 自定义溶剂参数（用于 custom 模型）
  eps?: number;           // 介电常数 ε
  eps_inf?: number;       // 光学介电常数 n²
  hbond_acidity?: number; // Abraham氢键酸度 α
  hbond_basicity?: number; // Abraham氢键碱度 β
  surface_tension?: number; // 表面张力 γ (cal/mol·Å²)
  carbon_aromaticity?: number; // 芳香碳原子比例 φ
  halogenicity?: number;  // 卤素原子比例 ψ
}

export interface DesolvationJobCreate {
  md_job_id: number;
  solvation_structure_id: number;
  method_level: string;
  desolvation_mode?: DesolvationMode;  // 默认为 'stepwise'
  solvent_config?: SolventConfig;      // 溶剂配置（可选，默认气相）
}

export interface LigandDesolvationResult {
  ligand_id: number;
  ligand_type: string;
  ligand_label: string;
  e_ligand: number;
  e_cluster_minus: number;
  delta_e: number;
}

export interface TypeSummary {
  ligand_type: string;
  avg_delta_e: number;
  std_delta_e: number;
  count: number;
  min_delta_e: number;
  max_delta_e: number;
}

export interface DesolvationEnergyResult {
  id: number;
  postprocess_job_id: number;
  solvation_structure_id: number;
  method_level: string;
  basis_set?: string;
  functional?: string;
  e_cluster: number;
  per_ligand_results: LigandDesolvationResult[];
  per_type_summary: TypeSummary[];
  created_at: string;
}

export interface QCProgress {
  total: number;
  completed: number;
  running: number;
  failed: number;
  progress_percent: number;
}

export interface DesolvationJobResponse {
  job_id: number;
  status: string;
  method_level: string;
  desolvation_mode: DesolvationMode;
  solvent_config?: SolventConfig;  // 溶剂模型配置
  created_at: string;
  started_at?: string;
  finished_at?: string;
  elapsed_seconds?: number;
  error_message?: string;
  result?: DesolvationEnergyResult;
  // 溯源信息
  solvation_structure_id?: number;
  composition_key?: string;  // 如 "Li-EC2-DMC1-PF6_1"
  md_job_id?: number;
  electrolyte_name?: string;
  qc_progress?: QCProgress;
}

export interface BatchDesolvationJobCreate {
  md_job_id: number;
  structure_ids: number[];
  method_level: string;
  desolvation_mode?: DesolvationMode;
  solvent_config?: SolventConfig;
  // Slurm 资源配置
  slurm_partition?: string;
  slurm_cpus?: number;
  slurm_time?: number;
}

export interface BatchDesolvationJobResponse {
  created_count: number;
  skipped_count: number;
  jobs: DesolvationJobResponse[];
}

export interface DesolvationOverviewResponse {
  md_job_id: number;
  electrolyte_name?: string;
  total_jobs: number;
  status_summary: Record<string, number>;
  jobs: DesolvationJobResponse[];
}

File: backend/app/schemas/desolvation.py
"""
Desolvation energy calculation schemas
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Literal
from datetime import datetime


class SolventConfigSchema(BaseModel):
    """溶剂配置（用于去溶剂化能计算的隐式溶剂模型）"""
    model: Literal["gas", "pcm", "smd", "custom"] = Field(default="gas", description="溶剂模型")
    solvent_name: Optional[str] = Field(default=None, description="预定义溶剂名称")
    # 自定义溶剂参数（SMD模型需要）
    eps: Optional[float] = Field(default=None, description="介电常数 ε")
    eps_inf: Optional[float] = Field(default=None, description="光学介电常数 n²")
    hbond_acidity: Optional[float] = Field(default=None, description="Abraham氢键酸度 α")
    hbond_basicity: Optional[float] = Field(default=None, description="Abraham氢键碱度 β")
    surface_tension: Optional[float] = Field(default=None, description="表面张力 γ (cal/mol·Å²)")
    carbon_aromaticity: Optional[float] = Field(default=None, description="芳香碳原子比例 φ")
    halogenicity: Optional[float] = Field(default=None, description="卤素原子比例 ψ")


class DesolvationJobCreate(BaseModel):
    """创建去溶剂化能任务"""
    md_job_id: int = Field(..., description="MD job ID")
    solvation_structure_id: int = Field(..., description="Solvation structure ID")
    method_level: str = Field(default="standard", description="Calculation method level: fast (6-31G(d)/B3LYP), standard (6-31++G(d,p)/B3LYP), accurate (6-311++G(2d,2p)/wB97XD)")
    desolvation_mode: Literal["stepwise", "full"] = Field(
        default="stepwise",
        description="Desolvation mode: stepwise (remove one ligand at a time) or full (remove all ligands at once)"
    )
    # 溶剂配置（可选，默认气相计算）
    solvent_config: Optional[SolventConfigSchema] = Field(
        default=None,
        description="Solvent configuration for implicit solvent model. If not provided, gas phase calculation is used."
    )


class LigandDesolvationResult(BaseModel):
    """单个配体的去溶剂化能结果"""
    ligand_id: int = Field(..., description="Ligand ID in the cluster")
    ligand_type: str = Field(..., description="Ligand type (e.g., FSI, EC, DMC)")
    ligand_label: str = Field(..., description="Ligand label (e.g., FSI_1, EC_2)")
    e_ligand: float = Field(..., description="Isolated ligand energy in A.U.")
    e_cluster_minus: float = Field(..., description="Cluster energy without this ligand in A.U.")
    delta_e: float = Field(..., description="Desolvation energy in kcal/mol")


class TypeSummary(BaseModel):
    """按类型汇总的统计"""
    ligand_type: str = Field(..., description="Ligand type")
    avg_delta_e: float = Field(..., description="Average desolvation energy in kcal/mol")
    std_delta_e: float = Field(..., description="Standard deviation in kcal/mol")
    count: int = Field(..., description="Number of ligands of this type")
    min_delta_e: float = Field(..., description="Minimum desolvation energy in kcal/mol")
    max_delta_e: float = Field(..., description="Maximum desolvation energy in kcal/mol")


class DesolvationEnergyResultSchema(BaseModel):
    """去溶剂化能结果"""
    id: int
    postprocess_job_id: int
    solvation_structure_id: int
    method_level: str
    basis_set: Optional[str] = None
    functional: Optional[str] = None
    e_cluster: float = Field(..., description="Complete cluster energy in A.U.")
    per_ligand_results: List[LigandDesolvationResult] = Field(default_factory=list)
    per_type_summary: List[TypeSummary] = Field(default_factory=list)
    created_at: datetime

    class Config:
        from_attributes = True


class DesolvationJobResponse(BaseModel):
    """去溶剂化能任务响应"""
    job_id: int
    status: str
    method_level: str
    desolvation_mode: str = "stepwise"  # stepwise or full
    solvent_config: Optional[SolventConfigSchema] = None  # 溶剂模型配置
    created_at: datetime
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    elapsed_seconds: Optional[float] = None
    error_message: Optional[str] = None
    result: Optional[DesolvationEnergyResultSchema] = None
    # 溯源信息
    solvation_structure_id: Optional[int] = None
    composition_key: Optional[str] = None  # 如 "Li-EC2-DMC1-PF6_1"
    md_job_id: Optional[int] = None
    electrolyte_name: Optional[str] = None  # 电解液配方名称
    qc_progress: Optional[Dict[str, Any]] = None  # QC任务进度 {"total": 10, "completed": 5, "running": 2}

    class Config:
        from_attributes = True


class BatchDesolvationJobCreate(BaseModel):
    """批量创建去溶剂化能任务"""
    md_job_id: int = Field(..., description="MD job ID")
    structure_ids: List[int] = Field(..., description="要计算的溶剂化结构ID列表")
    method_level: str = Field(default="standard", description="计算方法级别")
    desolvation_mode: Literal["stepwise", "full"] = Field(default="stepwise")
    solvent_config: Optional[SolventConfigSchema] = None
    # Slurm 资源配置
    slurm_partition: Optional[str] = Field(default="cpu", description="Slurm队列/分区")
    slurm_cpus: Optional[int] = Field(default=16, ge=1, le=64, description="CPU核心数")
    slurm_time: Optional[int] = Field(default=7200, ge=10, description="最大运行时间（分钟）")


class BatchDesolvationJobResponse(BaseModel):
    """批量创建响应"""
    created_count: int
    skipped_count: int  # 已存在的任务数
    jobs: List[DesolvationJobResponse]


class DesolvationOverviewResponse(BaseModel):
    """去溶剂化任务总览（用于监控面板）"""
    md_job_id: int
    electrolyte_name: Optional[str] = None
    total_jobs: int
    status_summary: Dict[str, int]  # {"COMPLETED": 3, "RUNNING": 2, "QUEUED": 1}
    jobs: List[DesolvationJobResponse]

